contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -556,11 +580,226 @@ pltsql_GetNewObjectId(VariableCache variableCache)
 
 	minOid = atooid(babelfish_dump_restore_min_oid);
 	Assert(OidIsValid(minOid));
-	if (ShmemVariableCache->nextOid >= minOid + 1)
+	if (variableCache->nextOid >= minOid + 1)
 		return;
 
-	ShmemVariableCache->nextOid = minOid + 1;
-	ShmemVariableCache->oidCount = 0;
+	variableCache->nextOid = minOid + 1;
+	variableCache->oidCount = 0;
+}
+
+static Oid
+pltsql_GetNewTempObjectId()
+{
+	Oid			result;
+	Oid			tempOidStart;
+	static Oid 	nextTempOid = InvalidOid;
+
+	/* safety check, we should never get this far in a HS standby */
+	if (RecoveryInProgress())
+		elog(ERROR, "cannot assign OIDs during recovery");
+
+	if (!OidIsValid(BUFFER_START_TO_OID)) /* InvalidOid means it needs assigning */
+	{
+		/* First check to see if another connection has already picked a start, then update. */
+		LWLockAcquire(OidGenLock, LW_EXCLUSIVE);
+		if (OidIsValid(ShmemVariableCache->tempOidStart))
+		{
+			temp_oid_buffer_start = OID_TO_BUFFER_START(ShmemVariableCache->tempOidStart);
+			nextTempOid = ShmemVariableCache->tempOidStart;
+		}
+		else
+		{
+			/* We need to pick a new start for the buffer range. */
+			tempOidStart = ShmemVariableCache->nextOid;
+
+			/*
+			 * Decrement ShmemVariableCache->oidCount to take into account the new buffer we're allocating
+			 */
+			if (ShmemVariableCache->oidCount < temp_oid_buffer_size)
```

Comment:
let us consider two cases:
1) oidCount is 0. It means that we do not have any remaining oid for current chunk. We come here, assign oids equal to buffer size and keep oid count as 0. However, we are not updating control file. If there is restart, both temp oid start and next oid will intersect. 
2) Similarly if oidCount is < buffer size, same problem could happen as we can go beyond the next oid in control file and a restart will lead to oid intersection among temp and next oid range

Reply:
What do you mean by control file? In the case you're giving here: 

- If oidCount is 0, then we are out of oids for the entire database. 

So first, let's say it's wraparound, ie ShmemVariableCache->nextOid = OID_MIN, since we will wrap around. If we create a temp table next, we will go through this code to pick a new tempOidStart. But tempOidStart will be pushed forward to FirstNormalObjectId by the code below. So the buffer will span from FirstNormalObjectId to FirstNormalObjectId + 65535. 

Then on line 625, we see that ShmemVariableCache->nextOid will be set to the end of the range. There will not be OID intersection. 

The other case to consider then that would be extraordinary is if ShmemVariableCache->nextOid = OID_MIN - 65535, so that we will now pick OID_MIN - 65535 as the beginning and OID_MIN as the end. That will bump ShmemVariableCache->nextOid to OID_MIN. The normal OID range will get bumped to FirstNormalObjectId by the code in GetNewObjectId, so there is no issue there as well.

Reply:
Discussed on call, we will add a write to control file for when oidCount = 0